modules/fedimint-wallet-server/src/lib.rs

```diff
@@ -1626,3 +1626,314 @@ mod tests {
+    use fedimint_testing::{prepare_snapshot, validate_migrations, BYTE_20, BYTE_32};
+    use fedimint_wallet_common::db::{
+        BlockHashKey, BlockHashKeyPrefix, DbKeyPrefix, PegOutBitcoinTransaction,
+        PegOutBitcoinTransactionPrefix, PegOutTxSignatureCI, PegOutTxSignatureCIPrefix,
+        PendingTransactionKey, PendingTransactionPrefixKey, RoundConsensusKey, UTXOKey,
+        UTXOPrefixKey, UnsignedTransactionKey, UnsignedTransactionPrefixKey,
+    };
+    use fedimint_wallet_common::{
+        PegOutFees, PendingTransaction, Rbf, RoundConsensus, SpendableUTXO, UnsignedTransaction,
+        WalletOutputOutcome,
+    };
+    use futures::StreamExt;
+    use rand::rngs::OsRng;
+    use secp256k1::Message;
+    use strum::IntoEnumIterator;
+
+    use crate::{Wallet, WalletGen};
+
+    /// Create a database with version 0 data. The database produced is not
+    /// intended to be real data or semantically correct. It is only
+    /// intended to provide coverage when reading the database
+    /// in future code versions. This function should not be updated when
+    /// database keys/values change - instead a new function should be added
+    /// that creates a new database backup that can be tested.
+    async fn create_db_with_v0_data(mut dbtx: DatabaseTransaction<'_>) {
+        dbtx.insert_new_entry(&BlockHashKey(BlockHash::from_slice(&BYTE_32).unwrap()), &())
+            .await;
+
+        let utxo = UTXOKey(bitcoin::OutPoint {
+            txid: Txid::from_slice(&BYTE_32).unwrap(),
+            vout: 0,
+        });
+        let spendable_utxo = SpendableUTXO {
+            tweak: BYTE_32,
+            amount: Amount::from_sat(10000),
+        };
+        dbtx.insert_new_entry(&utxo, &spendable_utxo).await;
+
+        let round_consensus = RoundConsensus {
+            block_height: 5000,
+            fee_rate: Feerate { sats_per_kvb: 1000 },
+            randomness_beacon: BYTE_32,
+        };
+        dbtx.insert_new_entry(&RoundConsensusKey, &round_consensus)
+            .await;
+
+        let unsigned_transaction_key = UnsignedTransactionKey(Txid::from_slice(&BYTE_32).unwrap());
+
+        let mut selected_utxos: Vec<(UTXOKey, SpendableUTXO)> = vec![];
+        selected_utxos.push((utxo.clone(), spendable_utxo));
+
+        let destination = Script::new_v0_p2wpkh(&WPubkeyHash::from_slice(&BYTE_20).unwrap());
+        let output: Vec<TxOut> = vec![TxOut {
+            value: 10000,
+            script_pubkey: destination.clone(),
+        }];
+
+        let transaction = Transaction {
+            version: 2,
+            lock_time: PackedLockTime::ZERO,
+            input: vec![TxIn {
+                previous_output: utxo.0,
+                script_sig: Default::default(),
+                sequence: Sequence::ENABLE_RBF_NO_LOCKTIME,
+                witness: bitcoin::Witness::new(),
+            }],
+            output,
+        };
+
+        let inputs = vec![Input {
```

Comment:
Hmm do you guys have thoughts on how we should handle persisted structs from external crates? I think there is a small chance that if we bump the version of the external crate, the structure has changed, but now it becomes difficult to write a db migration since the external crate likely doens't expose the old structure as well.

Reply:
Part of the reason we need migration tests. IMO - developers will have to import old versions of their dependencies (in parallel to new ones, possibly) for as long as they want to maintain migration window.

Reply:
The good thing is we use BTC consensus encoding here, which is unlikely to change. But agree with @dpc in the general case.